### ISSUE TITLE
Dreyfus is not detected in rebar scripts

### ISSUE THREAD

**Issue by willholley** [apache/couchdb#2265]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

When building CouchDB master I noticed that `ken` was not updating search indexes. Upon further investigation, it transpired that the conditional compilation flag `HAVE_DREYFUS` (see [here](https://github.com/apache/couchdb-ken/blob/tmp/rebar.config.script#L13)) is not set when the build is run in a clean environment. The impact of this is that background indexing for search indexes is disabled, though every other piece of search functionality seems to work fine.
 
I find that if I force `ken` to rebuild (e.g. delete the `ebin` directory from `src/ken` and run `make release`), the feature detection works!

## Steps to Reproduce

In a local CouchDB git repository, add debug output to rebar by adding `-vvv` to the `couch` target in the `Makefile`:

```
couch: config.erl
	@COUCHDB_VERSION=$(COUCHDB_VERSION) COUCHDB_GIT_SHA=$(COUCHDB_GIT_SHA) $(REBAR) -vvv compile $(COMPILE_OPTS)
	@cp src/couch/priv/couchjs bin/
```

Then run a clean build:

```
$ make clean
$ ./configure -c
$ make dist
```

Observe the debug output when compiling `ken`. `erl_opts` is what we're interested in:

```
...
==> ken (compile)
DEBUG: Matched required ERTS version: 8.3.5.6 -> .*
DEBUG: Matched required OTP release: 19 -> 19|20|21|22
DEBUG: Min OTP version unconfigured
DEBUG: erl_opts [debug_info,{i,"../"}]
...
```

If you then force `ken` to rebuild, you'll get the expected output:

```
$ rm -rf src/ken/ebin`
$ make dist

...
==> ken (compile)
DEBUG: Matched required ERTS version: 8.3.5 -> .*
DEBUG: Matched required OTP release: 19 -> 19|20|21|22
DEBUG: Min OTP version unconfigured
DEBUG: erl_opts [debug_info,{d,'HAVE_DREYFUS'},{i,"../"}]
...
```

I first detected this when building packages using the couchdb-ci Dockerfiles but can reproduce it locally using the above steps.

## Expected Behaviour

For CouchDB master, `HAVE_DREYFUS` should be set when `dreyfus` is present (which is now always?).

## Additional Context

`ken` isn't the only module to use the `HAVE_DREYFUS` compilation flag - Mango also uses it [here](https://github.com/apache/couchdb/blob/master/src/mango/rebar.config.script#L14). Knowing very little about rebar, my guess is that it's significant that `mango` is "part" of the CouchDB application whereas `ken` is an external dependency. At least, they are specified differently in the [rebar.config](https://github.com/apache/couchdb/blob/master/rebar.config.script#L91) that perhaps impacts compilation order / dependency resolution.

**Comment by willholley** [comment]
a seemingly critical bit from http://erlang.org/doc/man/code.html#lib_dir-1:
> If a regular directory called Name or Name-Vsn exists in the code path with an ebin subdirectory, the path to this directory is returned (not the ebin directory).

If I manually create `src/dreyfus/ebin` before running the build, `HAVE_DREYFUS` is set as expected.

**Comment by willholley** [comment]
This makes sense to me now. Our rebar configuration results in imported dependencies being compiled before any modules defined in the couchdb source tree. Historically, `dreyfus` and `ken` were peers; `dreyfus` would be built before `ken` and the feature detection in `ken` would work.

Now that `dreyfus` has been imported into the CouchDB source tree this is no longer true; `ken` will be compiled first as an external dependency and the detection of `dreyfus` fails.

Potentially we could import `ken` into the couchdb source tree in the same way as `dreyfus` so there are peers again. Alternatively we could just remove the conditional compilation flag given `dreyfus` now always exists in CouchDB.

**Comment by willholley** [comment]
fixed by #2266
